programs/sunrise_beam/src/state.rs

```diff
@@ -70,93 +83,82 @@ impl ControllerState {
         if let Some(bump) = input.new_gsol_mint_authority_bump {
             self.gsol_mint_authority_bump = bump;
         }
+        if let Some(window) = input.new_alloc_window {
+            self.alloc_window = window;
+        }
         if let Some(yield_account) = input.new_yield_account {
             self.yield_account = yield_account;
         }
     }
 
     /// A return value of [None] indicates no space was found.
-    pub fn push_allocation(&mut self, new_allocation: Allocation) -> Result<Option<()>> {
+    pub fn push_allocation(&mut self, new_allocation: BeamDetails) -> Result<Option<()>> {
         if self.contains_beam(&new_allocation.beam) {
-            return Err(BeamProgramError::DuplicateBeamEntry.into());
+            return Err(BeamError::DuplicateBeamEntry.into());
         }
 
-        let maybe_found = self
-            .allocations
-            .iter_mut()
-            .find(|x| x.beam == Pubkey::default());
-
-        if let Some(allocation) = maybe_found {
+        let found = self.get_mut_beam_details(&Pubkey::default());
+        if let Some(allocation) = found {
             *allocation = new_allocation;
 
             return Ok(Some(()));
         }
 
-        todo!("Modify allocations");
-
         Ok(None)
     }
 
     /// Returns the number of active beams.
     pub fn beam_count(&self) -> usize {
         self.allocations
             .iter()
-            .filter(|x| x.beam != Pubkey::default())
+            .filter(|x| **x != BeamDetails::default())
             .count()
     }
 
-    /// Returns [None] if the allocation is not present.
-    pub fn remove_beam(&mut self, beam: &Pubkey) -> Option<()> {
-        if !self.contains_beam(beam) {
-            return None;
-        }
-
-        let found = self
-            .allocations
-            .iter_mut()
-            .find(|x| x.beam == *beam)
-            .unwrap();
+    pub fn get_beam_details(&self, beam_key: &Pubkey) -> Option<&BeamDetails> {
+        self.allocations.iter().find(|x| x.beam == *beam_key)
+    }
 
-        let _allocation = found.allocation;
-        *found = Allocation::default();
+    pub fn get_mut_beam_details(&mut self, beam_key: &Pubkey) -> Option<&mut BeamDetails> {
+        self.allocations.iter_mut().find(|x| x.beam == *beam_key)
+    }
 
-        todo!("Draining mode logic and modify allocations");
+    /// Returns [None] if the allocation is not present.
```

Comment:
Should be an error?

Reply:
I chose to separate errors from the state's internal functions. 

The error is thrown in the instruction based on the following code:
```rust
if ctx.accounts.state.remove_beam(&beam).is_none() {
    return Err(BeamError::UnidentifiedBeam.into());
 }
```

Reply:
Ok